### PR TITLE
Add PMW3389 optical sensor Support (Updated)

### PR DESCRIPTION
**Pull request by Alabastard-64**
## Description
Update to PR #15514 as there were significant changes to this section of code on the development branch.
Adds support for PMW3389 sensor from pixart as an additional pointing device.
Much of the code is duplicated from PMW3360 code, this was done as a separate file to keep the changes to the registers clean (some where moved around and new ones added) and due to the change in CPI settings updates.

This code has been tested and works with the latest dev branch and a as of yet unreleased keyboard (waiting on this PR)

## Types of Changes

- [X] Core
- [ ] Bugfix
- [X] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [X] Documentation

## Issues Fixed or Closed by This PR

* Lack of proper PMW3389 support (it would not write CPI settings properly)

## Checklist

- [X] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [X] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [X] My change requires a change to the documentation.
- [X] I have updated the documentation accordingly.
- [X] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [X] I have added tests to cover my changes.
- [X] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by Alabastard-64** [comment]
Currently this was done as separate files but it is possible to add this to the pmw3360 driver with some ifdefs
and duplication of the registers section.

Let me know if it would be preferred to add this in that way (as it may be more maintainable if slightly messier)

**Comment by drashna** [comment]
Could you run a format pass on the files added/touched?  (`qmk format-c` will do the trick, despite the poor name)

**Comment by Alabastard-64** [comment]
no idea why the linter is not liking that file...
it is almost identical to pmw3360.c and I ran it through the qmk c formatter

**Review by drashna — APPROVED** [review]

File: drivers/sensors/pmw3389.c [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+    pmw3389_spi_start();
+    spi_write(REG_SROM_Load_Burst | 0x80);
+    wait_us(15);
+
+    unsigned char c;
+    for (int i = 0; i < FIRMWARE_LENGTH; i++) {
+        c = (unsigned char)pgm_read_byte(firmware_data + i);
+        spi_write(c);
+        wait_us(15);
+    }
```
Comment (drashna) [inline drivers/sensors/pmw3389.c:218]:
```suggestion
#ifdef PMW3389_LEGACY_FIRMWARE_UPLOAD
    unsigned char c;
    for (int i = 0; i < FIRMWARE_LENGTH; i++) {
        c = (unsigned char)pgm_read_byte(firmware_data + i);
        spi_write(c);
        wait_us(15);
    }
#else
    spi_transmit(firmware_data, sizeof(firmware_data));
#endif
```

File: drivers/sensors/pmw3360.h [hunk 2]
```diff
@@ -56,17 +56,6 @@
...
-/*
-The pmw33660 and pmw3389 use the same registers and timing and such.
-The only differences between the two is the firmware used, and the
-range for the DPI. So add a semi-secret hack to allow use of the
-pmw3389's firmware blob.  Also, can set the max cpi range too.
-This should work for the 3390 and 3391 too, in theory.
-*/
-#ifndef PMW3360_FIRMWARE_H
-#    define PMW3360_FIRMWARE_H "pmw3360_firmware.h"
-#endif
```
Comment (daskygit) [inline drivers/sensors/pmw3360.h:68]: Removing this requires pmw3360.c be updated.
https://github.com/qmk/qmk_firmware/blob/cfcbc089b77295bda68aba39ab0e516cf1f78a1e/drivers/sensors/pmw3360.c#L24
should now be
```c
#include "pmw3360_firmware.h"
```

**Review by daskygit — CHANGES_REQUESTED** [review]

File: drivers/sensors/pmw3360.h (hunk 2 above)
Comment (Alabastard-64) [reply drivers/sensors/pmw3360.h:68]: Ah yeah I missed that thanks.
I guess that part didn't technically need to be removed, but with this correction it does makes all the drivers that need to load firmware consistent.

**Review by daskygit — APPROVED** [review]

**Comment by daskygit** [comment]
Thanks, a slightly weak approval as I don't have hardware to test though.

**Comment by Alabastard-64** [comment]
Dang after testing it seems that the new initialization code doesn't seem to be working with the PMW3389 currently I need to enable PMW3389_LEGACY_FIRMWARE_UPLOAD in order to get the sensor working.
Not sure why it isn't working for the 3389 if it works for the 3360

**Comment by drashna** [comment]
odd, may be worth removing that change, then.  

There is some timing stuff for it, in the docs on the pmw3360, but in testing on multiple devices didn't trigger any issues.

**Comment by Alabastard-64** [comment]
hhmmm test a few times now and repeatably I get no response from the sensor if using spi_transmit to load the firmware.
Works totally fine using the old method though on both of the sensors I have on hand (I only have two PMW3389's and both are the same breakout board from Jack enterprises on tindie)
I'll submit another PR with the reversion and add a comment on it's failure during testing.

**Pull request merged**